### DISCUSSION TITLE
exited abnormally with code 127  [pylsp]

### DISCUSSION THREAD

**Discussion by CDitzel** [joaotavora/eglot#1262] [General]
I am connecting via Tramp to an aws remote machine. Having installed pylsp on both the remote and my local machine, without further configuration I am starting eglot and get 

```
[stderr] /bin/sh: 1: ~/.local/bin/pylsp: not found
[internal] Wed Aug  2 16:06:12 2023:
(:message "Connection state changed" :change
		  "exited abnormally with code 127\n")
```

I checked that pylsp is in fact in the path on both machines and also located under ~/.local/bin/pylsp so I am a bit confused why eglot cannot find it. 

Can someone help here please?

**Comment by triHims** [comment]
I feel this is a problem of emacs path. 
is this a mac machine?

**Reply by CDitzel** [reply]
no linux (ubuntu 22.04). Funny enough, when I ssh into remote and try eglot in a python file over there, it works. Also, when I edit a file locally with my local emacs instance it works. Only when I tramp into the remote, it cant find pylsp on path.  so I dont understand whats going on really

**Comment by iquiw** [comment]
Try to configure `tramp-remote-path`.

```emacs-lisp
(add-to-list 'tramp-remote-path "~/.local/bin")
```

or add `tramp-own-remote-path` to `tramp-remote-path`.

ref. [Remote programs (TRAMP 2.6.0.29.1 User Manual)](https://www.gnu.org/software/emacs/manual/html_node/tramp/Remote-programs.html)

**Reply by CDitzel** [reply]
will try that, thank you. I am wondering, tramp does copy files to the local machine and transmits them again to the remote one upon saving. From what I can tell, the local machines emacs config is the one used for that and (if even existing) the remote machines config is not involved at all, is that understanding correct?

In this context, does clang, pylsp etc. needs to be installed on the local or the remote machine then? I have them on both but pylsp still seems to struggle to be found oO

**Reply by joaotavora** [reply]
> In this context, does clang, pylsp etc. needs to be installed on the local or the remote machine then?

Remote machine, in your case.  I.e. the language server must live in the same machine where the files live.  And the tip by @iquiw is spot on (adding `tramp-own-remote-path` to `tramp-remote-path`).  See TRAMP manual for details.

**Reply by joaotavora** [reply]
> From what I can tell, the local machines emacs config is the one used for that and (if even existing) the remote machines config is not involved at all, is that understanding correct?

Yes, this is correct.
